### PR TITLE
Inheritance upgrade / Add more features 

### PR DESCRIPTION
**Pull request by B-Stefan**
Enhancements for the inheritance feature. 

This code will add the features for: 

* Nested schema inheritance 
* Array Items inheritance
* @modelOptions inheritance 
* Multi-level inheritance

### CODE REVIEW

**Comment by B-Stefan** [comment]
This merge request ist not ready to merge. 
There is a bug with the connection options for extended classes. 

Please don't merge this into the master at this point.

**Comment by B-Stefan** [comment]
Solved the bug - was just a configuration problem on our site.

File: src/prop.ts [hunk 1]
```diff
@@ -218,12 +218,15 @@ function baseProp(
   }
 
   if (whatis === WhatIsIt.ARRAY) {
+
+    const virtualSchemaItem = _buildSchema(Type,  { _id: rawOptions._id });
```
Comment (hasezoey) [inline src/prop.ts]: why is this change needed?

File: src/typegoose.ts [hunk 2]
```diff
@@ -114,22 +114,14 @@ export function buildSchema<T, U extends NoParamConstructor<T>>(cl: U) {
...
-    // set next parent
-    parentCtor = Object.getPrototypeOf(parentCtor.prototype).constructor;
-  }
-  // get schema of current model
-  sch = _buildSchema(cl, sch);
+  const {schemaOptions = {}}: IModelOptions = Reflect.getMetadata(DecoratorKeys.ModelOptions, cl) || {};
+
+  const sch = _buildSchema(cl, schemaOptions);
+
+  Object.entries(schemaOptions)
```
Comment (hasezoey) [inline src/typegoose.ts]: why does it need to be applied after the schema is build? isnt it already applied in `_buildSchema`?

File: test/models/inheritanceClass.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import { arrayProp, prop } from '../../src/prop';
+import { getModelForClass, modelOptions } from '../../src/typegoose';
+
+@modelOptions({
+  schemaOptions: {
+    discriminatorKey: 'width',
```
Comment (hasezoey) [inline test/models/inheritanceClass.ts]: this is as far as i can tell not a valid schema Option ( https://mongoosejs.com/docs/guide.html#collection )

File: src/typegoose.ts (hunk 2 above)
Comment (hasezoey) [inline src/typegoose.ts]: why does it need to be applied after the schema is build? isnt it already applied in _buildSchema?

File: src/prop.ts (hunk 1 above)
Comment (hasezoey) [inline src/prop.ts]: why is this change needed?

File: test/models/inheritanceClass.ts (hunk 3 above)
Comment (B-Stefan) [reply test/models/inheritanceClass.ts]: It seems that the documentation is not accurate. I found an example of the option: 
https://mongoosejs.com/docs/api.html#singlenestedpath_SingleNestedPath-discriminator

And the typings of mongoose also allow this option.

Comment (hasezoey) [reply test/models/inheritanceClass.ts]: ok, thanks, could you now say me why you would want a discriminator in a nested path?

File: src/typegoose.ts (hunk 2 above)
Comment (B-Stefan) [reply src/typegoose.ts]: This is needed because the @prop decorators are executed before the getModelForClass method. This enhancement is covered by the test https://github.com/hasezoey/typegoose/pull/21/files#diff-ea520afea15a8887a0395e188d12c020R16


The root cause which makes this code necessary is in the schema.ts file (https://github.com/B-Stefan/typegoose/blob/r6/inheritance-upgrade/src/internal/schema.ts#L62)

The current execution order looks like this: 

1. Evaluation of @prop decorators 
2. One @prop decorator type is a `Garage` class 
3. It will hit the line (https://github.com/B-Stefan/typegoose/blob/r6/inheritance-upgrade/src/internal/schema.ts#L62) and create a new schema (without the options)
4. Evaluation of @modelOptions decorators will set metadata for the schema options
5. At some point somebody calls `buildSchema(Garage)`
6. It won't create a new Schema (https://github.com/B-Stefan/typegoose/blob/r6/inheritance-upgrade/src/internal/schema.ts#L62) because it already exists.
7. Options will be not set to the class.

File: src/prop.ts (hunk 1 above)
Comment (B-Stefan) [reply src/prop.ts]: To support Inheritance for items in arrays. This line is necessary to pass the test: https://github.com/hasezoey/typegoose/pull/21/files#diff-ea520afea15a8887a0395e188d12c020R38

File: src/typegoose.ts (hunk 2 above)
Comment (hasezoey) [reply src/typegoose.ts]: ok i think i now understood what you mean here, would it be better if we cache calls to prop & execture them when buildSchema is called?

File: test/models/inheritanceClass.ts (hunk 3 above)
Comment (hasezoey) [reply test/models/inheritanceClass.ts]: i think i understood it now, [thanks to this blog post](https://thecodebarbarian.com/mongoose-4.12-single-embedded-discriminators)

File: src/internal/schema.ts [hunk 4]
```diff
@@ -4,17 +4,42 @@ import { IModelOptions } from '../typegoose';
 import { EmptyVoidFn, NoParamConstructor } from '../types';
 import { DecoratorKeys } from './constants';
 import { buildSchemas, hooks, plugins, schemas, virtuals } from './data';
+import { getParentClasses } from './utils';
 
 /**
  * Private schema builder out of class props
```
Comment (hasezoey) [inline src/internal/schema.ts]: Could you make tsdoc for the new _buildSchema & getSchemaForClass differ from each other (to better distinguish them)

File: test/models/inheritanceClass.ts (hunk 3 above)
Comment (hasezoey) [reply test/models/inheritanceClass.ts]: PS: i understood it now, but why is this needed in the tests? because i cant find anywhere where a discriminator call is used

File: src/typegoose.ts (hunk 2 above)
Comment (B-Stefan) [reply src/typegoose.ts]: @hasezoey  Good idea! Thought about this as well and I think this will solve the problem. 
But on the other hand, it would add increase the complexity of the schema build process. 
Should we create some kind of improvement issue for this?

Comment (hasezoey) [reply src/typegoose.ts]: yes, would be good, and i think too much for v6.0.0

**Comment by B-Stefan** [comment]
Sure, I will rebase this and make all requested changes tomorrow.
Just a question: Is there any style guide beyond the tslint rules? If I would know the style I could solve the style issues aswell.

**Comment by hasezoey** [comment]
Current style "guides":
- tsc (`npm run build`)
- tslint (with all the settings, we might migrate over to eslint when it gets better typescript integration)
  - some rules can sometimes somehow not be applied, or are not existing yet in tslint.json
- vscode
- [CONTRIBUTING](https://github.com/hasezoey/typegoose/blob/r6/master/.github/CONTRIBUTING.md)

**Comment by hasezoey** [comment]
*Update to https://github.com/hasezoey/typegoose/pull/21#issuecomment-528139782*

in c7c979cbd000057da43e5ec9d0810fc717837905 i added the rule, that was most concerning me in this pr, just run `npm lint` to see all style-errors and `npm lint -- --fix` to automaticly fix them

**Comment by hasezoey** [comment]
I think we might need to implement the decorator cache in v6, because "i" found the first victim of the issue (@fabioformosa) with https://github.com/fabioformosa/typegoose-validation-error

**Comment by hasezoey** [comment]
for what is 7cc628adb1f2ba7350d187dc1e1959fa510dd4f7 necessary?

**Comment by B-Stefan** [comment]
If you desing a class like this: 

```typescript 
class Parent {
  @prop()
  public type: string
}
@modelOptions({
   schemaOptions: {
       discriminatorKey: "type"
   }
})
class Child extends Parent {
   @prop
  public someOther: string
}

```

if you call `getDiscriminatorModelForClass(Child)` you will get an  error like "The class can't override the discriminatorKey property"

**Comment by hasezoey** [comment]
im confused, i will trust you for now, because i still have to look at #23

**Comment by hasezoey** [comment]
Question: what did you do to the commits? (a rebase | a merge?)

**Comment by B-Stefan** [comment]
Tried to rebase but I think I messed up :D 
Will look into this to solve this and get a clean state

**Comment by B-Stefan** [comment]
I have a clean local state for this branch but I have to figure out why some tests fail.

File: src/internal/schema.ts (hunk 4 above)
Comment (hasezoey) [reply src/internal/schema.ts]: fixed in e3bea3322cb7ae97825578667a1b4f39eb12c492

**Comment by B-Stefan** [comment]
Found the reason why tests are failing: https://github.com/hasezoey/typegoose/blob/r6/master/src/internal/utils.ts#L165 need the metadata of the class level. Tests should pass after #23  is finished.

**Comment by hasezoey** [comment]
im still confused about #23 (& #24), lets see how it will turn out

**Comment by hasezoey** [comment]
@B-Stefan could you rebase this pr onto latest `r6/master` again?

**Comment by B-Stefan** [comment]
@hasezoey  I merged all pending merge-request on Friday together (only local). It worked as expect. 

I will rebase this merge-request as soon as #24 got merged into the r6/master.

**Comment by hasezoey** [comment]
@B-Stefan because #24 got merged, could you update this PR?

**Comment by B-Stefan** [comment]
@hasezoey Thanks for the hint! I will update this PR on Monday

**Comment by B-Stefan** [comment]
@hasezoey  Thanks for all the work you did! 
Most of my code is not necessary anymore or you implemented a similar solution! 👍

File: src/prop.ts [hunk 5]
```diff
@@ -245,13 +245,13 @@ function baseProp(
...
+          _id: typeof rawOptions._id === 'boolean' ? rawOptions._id : true
+        });
         schemas.get(name)[key] = {
           ...schemas.get(name)[key][0], // [0] is needed, because "initasArray" adds this (empty)
           ...options,
-          type: [{
-            ...(typeof options._id === 'boolean' ? { _id: options._id } : {}),
-            ...subSchema
-          }]
+          type: [virtualSchemaArrayItem]
```
Comment (hasezoey) [inline src/prop.ts]: what does this change improve?

File: src/typegoose.ts [hunk 6]
```diff
@@ -199,8 +199,13 @@ export function getDiscriminatorModelForClass<T, U extends AnyParamConstructor<T
   if (models.get(name)) {
     return models.get(name) as ReturnModelType<U, T>;
   }
+  const sch = buildSchema(cl) as mongoose.Schema & {paths: object};
+
+  const discriminatorKey =  sch.get('discriminatorKey');
+  if (sch.path(discriminatorKey)) {
+    sch.paths[discriminatorKey].options.$skipDiscriminatorCheck = true;
```
Comment (hasezoey) [inline src/typegoose.ts]: *i dont fully understand mongoose, thats why i ask this*

what does "$skipDiscriminatorCheck" do?

File: src/prop.ts (hunk 5 above)
Comment (B-Stefan) [reply src/prop.ts]: I think just the readability. It will not change the behavior because I pass `(typeof options._id === 'boolean' ? { _id: options._id } : {}` int `buildModel`

File: src/typegoose.ts (hunk 6 above)
Comment (B-Stefan) [reply src/typegoose.ts]: It is a little bit hacky because it is an internal option normally not used outside of the mongoose lib. But it allows you to override the key of a discriminator class.  This is necessary to support custom discriminatorKeys.

Comment (hasezoey) [reply src/typegoose.ts]: after some testing, i couldnt see a difference in the tests, so why was this (whole (in this function)) change needed?

Comment (B-Stefan) [reply src/typegoose.ts]: I modified the discriminator class to cover this edge case (https://github.com/typegoose/typegoose/pull/21/files#diff-5ea243b73204be263b19204e1f63a933R18) if you remove this line you will see that one test is failing.

Comment (hasezoey) [reply src/typegoose.ts]: i mean it worked before, it had the correct discriminator (in the discriminators.ts file), but your *new* test `should set all parent props for nested array items` would fail

Comment (B-Stefan) [reply src/typegoose.ts]: Okay thanks, for all the testing effort! We are using the discriminator feature in our application heavily and I hade some trouble to support custom discriminator keys. This was my solution to support more complex structures where the discriminator key is a different property than `__t`.  I'm open to any suggestions and enhancements! :)

**Comment by hasezoey** [comment]
could you do some documentation for the "new" discriminator? (because of the new "default" i see on a discriminator porp)

**Comment by B-Stefan** [comment]
@hasezoey  Do you mean documentation for the test class (https://github.com/typegoose/typegoose/pull/21/files#diff-5ea243b73204be263b19204e1f63a933R7)?

If so I would like to modify the test class to make it more explicit what you can do with this feature.
With this change, we support more complex cases but don't change any current behavior.

**Comment by hasezoey** [comment]
@B-Stefan sorry if i was not clear enough, i meant documentation [on prop](https://github.com/typegoose/typegoose/blob/r6/master/github-page/_docs/decorators/prop.md) and/or make an advanced guide for this in [this folder](https://github.com/typegoose/typegoose/tree/r6/master/github-page/_guides/advanced)

File: src/prop.ts [hunk 7]
```diff
@@ -245,13 +245,13 @@ function baseProp(
 
     switch (whatis) {
       case WhatIsIt.ARRAY:
+        const virtualSchemaArrayItem = buildSchema(Type, {
```
Comment (hasezoey) [inline src/prop.ts]: could you combine this buildSchema call with the `case WhatIsIt.NONE` buildSchema call? (and eventually add it into `case WhatIsIt.MAP`)

Comment (hasezoey) [reply src/prop.ts]: i will try to do this later

**Comment by hasezoey** [comment]
i will merge it for now, but if you could, could you do some documentation as described above for discriminator change?
